src/functions.sh

```diff
@@ -71,18 +71,36 @@ ipt() {
     ip6tables $* >> ${OUTPUT_TARGET} 2>&1
 }
 
+
 # wrapper to call tc to allow debug logging
 tc_wrapper() {
-    tc_args=$*
     sqm_trace "${TC_BINARY} $*"
-    ${TC_BINARY} $* >> ${OUTPUT_TARGET} 2>&1
+    local LAST_ERROR=$( ${TC_BINARY} $* 2>&1 )
+    #sqm_error "LAST_ERROR:: ${LAST_ERROR}"
+    echo ${LAST_ERROR} >> ${OUTPUT_TARGET} 2>&1
+    if [ -z "$LAST_ERROR" ] ; then
```

Comment:
I don't like relying on empty output for error / no error. This works fine on my openwrt:

```
OUTPUT=$(command)
RET=$?
```
So why can't we just do that?

Reply:
Because it fails on my openwrt...

Reply:
Honestly I would love making this dependent on the exit status, but in my testing this does not trigger even when the output is non-empty.
Now that might be a bug in busybox that needs fixing, but at least right now this seems to be no robust solution, unfortunately.

Reply:
# wrapper to call tc to allow debug logging
tc_wrapper() {
    sqm_trace "${TC_BINARY} $*"
    local LAST_ERROR=$( ${TC_BINARY} $* 2>&1 )
    #sqm_error "LAST_ERROR:: ${LAST_ERROR}"
    echo ${LAST_ERROR} >> ${OUTPUT_TARGET} 2>&1
    if [ -z "$LAST_ERROR" ] ; then
        sqm_debug "tc_wrapper: SUCCESS: ${TC_BINARY} $*"
    else                     
        # this went south, try to capture & report more detail
        sqm_error "tc_wrapper: FAILURE: ${TC_BINARY} $*"
        sqm_error "tc_wrapper: LAST ERROR: ${LAST_ERROR}"
    fi
}

With an invalid keyword in iqdisc_opts returns:

root@nacktmulle:~# SQM_DEBUG=1 SQM_VERBOSITY_MAX=8 /etc/init.d/sqm stop ; SQM_DEBUG=1 SQM_VERBOSITY_MAX=8 /etc/init.d/sqm start
SQM: Stopping SQM on pppoe-wan
SQM: ifb associated with interface pppoe-wan: ifb4pppoe-wan
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev pppoe-wan ingress
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev pppoe-wan root
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev ifb4pppoe-wan root
SQM: /usr/lib/sqm/stop-sqm: ifb4pppoe-wan shaper deleted
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev ifb4pppoe-wan down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete ifb4pppoe-wan type ifb
SQM: /usr/lib/sqm/stop-sqm: ifb4pppoe-wan interface deleted
SQM: Starting SQM script: layer_cake.qos on pppoe-wan, in: 46246 Kbps, out: 9890 Kbps
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc replace dev TMP_IFB_4_SQM root cake
SQM: QDISC cake is useable.
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev TMP_IFB_4_SQM down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete TMP_IFB_4_SQM type ifb
SQM: Starting layer_cake.qos
SQM: ifb associated with interface pppoe-wan: 
SQM: Currently no ifb is associated with pppoe-wan, this is normal during starting of the sqm system.
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name ifb4pppoe-wan type ifb
SQM: LLA: default link layer adjustment method for cake is cake
SQM: cake link layer adjustments:  overhead 34 mpu 64
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc add dev pppoe-wan root cake bandwidth 9890kbit overhead 34 mpu 64 diffserv3 nat dual-srchost
SQM: egress shaping activated
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc replace dev TMP_IFB_4_SQM ingress
SQM: QDISC ingress is useable.
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev TMP_IFB_4_SQM down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc add dev pppoe-wan handle ffff: ingress
SQM: LLA: default link layer adjustment method for cake is cake
SQM: cake link layer adjustments:  overhead 34 mpu 64
SQM: ERROR: tc_wrapper: FAILURE: /usr/sbin/tc qdisc add dev ifb4pppoe-wan root cake bandwidth 46246kbit overhead 34 mpu 64 diffserv3 nat dual-dsthost ingress2
SQM: ERROR: tc_wrapper: LAST ERROR: What is "ingress2"?
Usage: ... cake [ bandwidth RATE | unlimited* | autorate-ingress ]
                [ rtt TIME | datacentre | lan | metro | regional |
                  internet* | oceanic | satellite | interplanetary ]
                [ besteffort | diffserv8 | diffserv4 | diffserv3* ]
                [ flowblind | srchost | dsthost | hosts | flows |
                  dual-srchost | dual-dsthost | triple-isolate* ]
                [ nat | nonat* ]
                [ wash | nowash* ]
                [ split-gso* | no-split-gso ]
                [ ack-filter | ack-filter-aggressive | no-ack-filter* ]
                [ memlimit LIMIT ]
                [ ptm | atm | noatm* ] [ overhead N | conservative | raw* ]
                [ mpu N ] [ ingress | egress* ]
                (* marks defaults)
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev ifb4pppoe-wan up
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc filter add dev pppoe-wan parent ffff: protocol all prio 10 u32 match u32 0 0 flowid 1:1 action mirred egress redirect dev ifb4pppoe-wan
SQM: ingress shaping activated
SQM: layer_cake.qos was started on pppoe-wan successfully


But the following:
# wrapper to call tc to allow debug logging
tc_wrapper() {
    sqm_trace "${TC_BINARY} $*"
    local LAST_ERROR=$( ${TC_BINARY} $* 2>&1 )
    RET=$?
    #sqm_error "LAST_ERROR:: ${LAST_ERROR}"
    echo ${LAST_ERROR} >> ${OUTPUT_TARGET} 2>&1    

                             
    if [ "$RET" = "0" ] ; then
        sqm_debug "tc_wrapper: SUCCESS: ${TC_BINARY} $*"
    else                     
        # this went south, try to capture & report more detail
        sqm_error "tc_wrapper: FAILURE: ${TC_BINARY} $*"
        sqm_error "tc_wrapper: LAST ERROR: ${LAST_ERROR}"
    fi
}


results in:

root@nacktmulle:~# SQM_DEBUG=1 SQM_VERBOSITY_MAX=8 /etc/init.d/sqm stop ; SQM_DEBUG=1 SQM_VERBOSITY_MAX=8 /etc/init.d/sqm start
SQM: Stopping SQM on pppoe-wan
SQM: ifb associated with interface pppoe-wan: ifb4pppoe-wan
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev pppoe-wan ingress
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev pppoe-wan root
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc del dev ifb4pppoe-wan root
SQM: /usr/lib/sqm/stop-sqm: ifb4pppoe-wan shaper deleted
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev ifb4pppoe-wan down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete ifb4pppoe-wan type ifb
SQM: /usr/lib/sqm/stop-sqm: ifb4pppoe-wan interface deleted
SQM: Starting SQM script: layer_cake.qos on pppoe-wan, in: 46246 Kbps, out: 9890 Kbps
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc replace dev TMP_IFB_4_SQM root cake
SQM: QDISC cake is useable.
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev TMP_IFB_4_SQM down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete TMP_IFB_4_SQM type ifb
SQM: Starting layer_cake.qos
SQM: ifb associated with interface pppoe-wan: 
SQM: Currently no ifb is associated with pppoe-wan, this is normal during starting of the sqm system.
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name ifb4pppoe-wan type ifb
SQM: LLA: default link layer adjustment method for cake is cake
SQM: cake link layer adjustments:  overhead 34 mpu 64
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc add dev pppoe-wan root cake bandwidth 9890kbit overhead 34 mpu 64 diffserv3 nat dual-srchost
SQM: egress shaping activated
SQM: ip_wrapper: SUCCESS: /sbin/ip link add name TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc replace dev TMP_IFB_4_SQM ingress
SQM: QDISC ingress is useable.
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev TMP_IFB_4_SQM down
SQM: ip_wrapper: SUCCESS: /sbin/ip link delete TMP_IFB_4_SQM type ifb
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc add dev pppoe-wan handle ffff: ingress
SQM: LLA: default link layer adjustment method for cake is cake
SQM: cake link layer adjustments:  overhead 34 mpu 64
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc qdisc add dev ifb4pppoe-wan root cake bandwidth 46246kbit overhead 34 mpu 64 diffserv3 nat dual-dsthost ingress2
SQM: ip_wrapper: SUCCESS: /sbin/ip link set dev ifb4pppoe-wan up
SQM: tc_wrapper: SUCCESS: /usr/sbin/tc filter add dev pppoe-wan parent ffff: protocol all prio 10 u32 match u32 0 0 flowid 1:1 action mirred egress redirect dev ifb4pppoe-wan
SQM: ingress shaping activated
SQM: layer_cake.qos was started on pppoe-wan successfully

Reply:
What is the exact code that is not working? Because there are bugs in several of the various versions above.

Specifically, the return code should be saved immediately after the command, before anything else... Putting an 'echo' in-between overrides the return code.

So this:

```
 local LAST_ERROR=$( ${TC_BINARY} $* 2>&1 )
    #sqm_error "LAST_ERROR:: ${LAST_ERROR}"
    echo ${LAST_ERROR} >> ${OUTPUT_TARGET} 2>&1
    if [ $? -neq 0 ] ; then
```

won't work; but this:

```
    local LAST_ERROR=$( ${TC_BINARY} $* 2>&1 )
    res=$?
    #sqm_error "LAST_ERROR:: ${LAST_ERROR}"
    echo ${LAST_ERROR} >> ${OUTPUT_TARGET} 2>&1
    if [ $res -neq 0 ] ; then
```

should... And if it doesn't, I'd like to understand why before working around it :)